nf15 (#121)
blues/app-accelerators: 15-greenhouse-monitor/README.md

## Notehub

Sign up for a free account on [notehub.io](https://notehub.io) and [create a new project](https://dev.blues.io/quickstart/notecard-quickstart/notecard-and-notecarrier-pi/#set-up-notehub).

This solution can be used to monitor a single greenhouse, or monitor multiple greenhouses at a facility by deploying multiple instances of this app. When monitoring multiple greenhouses at a facility, it can be useful to group the monitors at a facility into a Fleet. For more details, see [The Fleet Administrator's Guide](https://dev.blues.io/guides-and-tutorials/fleet-admin-guide/).

## Application Firmware

The application firmware is found under the [firmware](./firmware/) folder and can be built and uploaded to Swan using these development environments:

* PlatformIO extension for Visual Studio Code
* Arduino extension for Visual Studio Code
* Arduino IDE

We recommend using the PlatformIO extension for Visual Studio Code, as this is the easiest to set up and use, and provides a comprehensive development experience. However, if you're familiar with the Arduino IDE, that can be used as well but requires a little more setup.

### PlatformIO extension for VS Code

> There is no special setup required for the project beyond what is normally required to configure a PlatformIO project in VSCode. [This tutorial](https://dev.blues.io/quickstart/swan-quickstart/#using-the-vs-code-platformio-extension) explains how to install and use the PlatformIO.

Reviewer: "use the PlatformIO." -> "use PlatformIO."

* Blues Wireless Notecard
* Adafruit Seesaw
* Adafruit BusIO (dependency of Adafruit seesaw)
* Adafruit BME280 sensor
* TaskScheduler

#### Arduino IDE - Compiling/Uploading

To compile and upload the app firmware, open the sketch at [`firmware/greenhouse/greenhouse.ino`](firmware/greenhouse/greenhouse.ino) in the Arduino IDE.

### Configuring the ProductUID

Before uploading the app firmware, you should set the `ProductUID` to correspond to the project you created in [Notehub setup](#notehub) earlier.

There are two ways to configure the `ProductUID`, either using the in-browser terminal to send a request to the Notecard, or by editing the firmware source code. For more details on what the `ProductUID` is and how it set it please see [this guide](https://dev.blues.io/notehub/notehub-walkthrough/#finding-a-productuid).

#### Using the In-browser terminal

1. Connect the Notecarrier to your computer using a micro-USB cable.
2. Launch the in-browser terminal at [blues.dev](https://dev.blues.io/notecard-playground/)
3. Click the "USB Notecard" button under "Connect a Notecard".
4. Select the Notecard to connect to and click "Connect".

> 5. The terminal will display the firmware version of Notecard.

Reviewer: "of Notecard" -> "of the Notecard"

The app is configured using environment variables to set

* how often environment variables are checked,
* how often the sensors are read and checked for alerts, and
* how often monitoring events are sent.

These variables control the overall app behavior:

* `environment_update_mins` How often (in minutes) to check for environment variable updates from the Notecard. When not set, the default value is 5 minutes.

* `monitor_secs`: How often (in seconds) to read the current values from the sensors and check for alerts. When not set, the default value is 15 seconds.

* `report_mins` How often (in minutes) to report the current values from sensors to `greenhouse.qo`. When not set, the default value is 5 minutes.

You set these environment variables in Notehub. See our tutorial [Understanding Environment Variables](https://dev.blues.io/guides-and-tutorials/notecard-guides/understanding-environment-variables/) for a fuller description of how to set environment variables in Notehub.

### Sensor Names

The app senses and reports these values:

* `air_temp`: the air temperature in Celsius

* `air_humidity`: the relative humidity as a percent

* `air_pressure`: the air pressure in kPa

> * `soil_moisture`: an indication of how much moisture is in the soil. A higher value means more moisture, a lower value means less moisture.

Reviewer: This is kind of vague. If that's just a limitation of the sensor, though, that makes sense. But knowing the scale at least would help, I think, even if we can't know the unit of measurement. Same goes for `light_level`.

Reviewer: I now see you go over this in greater detail in the "Soil Moisture and Light Level" section. Maybe just adding a link to that part of the document here would be good, in that case.

  * No alert is produced when the soil temperature is below 40.

  * A warning alert is produced when the soil temperature is between 40 and 50.

  * A critical alert is produced when the soil temperature is above 50.

### Alert Events

When the app detects an alert condition, it captures details of the alert and sends an event to `alert.qo`. The event includes the overall alert level (warning/critical) and details of the sensor readings, similar to this:

```json
{
    "alert": "warning",
    "alert_seq": "first",
    "app": "nf15",
    "air_humidity": {
        "status": "ok",
        "value": 34.740234375
    },
    "air_pressure": {
        "status": "ok",
        "value": 1016.686328125
    },
    "air_temp": {
        "status": "ok",
        "value": 23.46999931335449
    },
    "light_level": {
        "alert": "warning",
        "status": "high",
        "value": 279
    },
    "soil_moisture": {
        "status": "ok",
        "value": 342
    },
    "soil_temp": {
        "status": "ok",
        "value": 26.38705444335938
    }
}
```

The event has these properties:

> * `alert`: Describes the overall alert level. This is the combined alert level of all of the sensors. Values are `warning`, `critical` or not present, meaning no alert, which happens when sensor readings return to normal.

Reviewer: > This is the combined alert level of all of the sensors.

Is it the highest level present between all the alerts? If so, I would state that explicitly.

  * [Qwiic 4-pin JST connector cable](https://www.adafruit.com/product/4399)
  * 1k&Omega; resistor
  * 3 [male-to-male jumper cables](https://www.adafruit.com/product/1957)
  * [half size breadboard](https://www.adafruit.com/product/4539)

## Hardware Setup

1. Assemble the Notecard, Notecarrier and antenna as described in our [quickstart tutorial](https://dev.blues.io/quickstart/notecard-quickstart/notecard-and-notecarrier-f/).

2. Connect the I2C Qwiic cable between the Notecarrier and the BME280 breakout board:

    1. Insert the Qwiic JST cable connector into one of the `F_I2C` connectors on the edge of the Notecarrier-F next to the USB port. You may also connect it to the I2C connector on the Swan.

    2. Insert the other end of the Qwiic JST cable into one of the JST connectors on the BM280 breakout board.

3. Connect the photo transistor:

   1. Connect the 3 jumper wires between the Notecarrier and breadboard as follows
      * `GND` on Notecarrier to the negative rail at the edge of the breadboard
      * `VMAIN` on Notecarrier to the positive power rail at the edge the breadboard
      * `A1` on Notecarrier to a row of points on the breadboard

>     2. Insert one leg of the 1k&Omega; resistor into the ground rail and the other leg to the same row of points connected to `A1` in last step above.

Author: I added the comma there and also in the line below, but I am curious why it is needed.

>     2. Insert one leg of the 1k&Omega; resistor into the ground rail and the other leg to the same row of points connected to `A1` in last step above.

Reviewer 2: It’s subjective so just a suggestion. I personally find it easier to read because it breaks up two different actions I need to take as a reader.

    3. Insert the longer leg of the photo transistor into the positive power rail and the shorter leg into the same row of points as the A1 wire.

>     ![](./assets/phototransistor.png)

Author: Thanks, I changed it after my A0 header stopped holding the cable firmly.  Will update.

### Configuring the ProductUID

Before uploading the app firmware, you should set the `ProductUID` to correspond to the project you created in [Notehub setup](#notehub) earlier.

There are two ways to configure the `ProductUID`, either using the in-browser terminal to send a request to the Notecard, or by editing the firmware source code. For more details on what the `ProductUID` is and how it set it please see [this guide](https://dev.blues.io/notehub/notehub-walkthrough/#finding-a-productuid).

#### Using the In-browser terminal

1. Connect the Notecarrier to your computer using a micro-USB cable.
2. Launch the in-browser terminal at [blues.dev](https://dev.blues.io/notecard-playground/)
3. Click the "USB Notecard" button under "Connect a Notecard".
4. Select the Notecard to connect to and click "Connect".
5. The terminal will display the firmware version of Notecard.
6. You can now enter a request to set the ProductUID and Serial Number of the device.

```json
{"req":"hub.set", "product":"<your-productUID-from-notehub>", "sn":"greenhouse-monitor"}
```

> You can also omit the serial number use Notehub to set it:

Reviewer 2: ```suggestion
You can also omit the serial number and use Notehub to set it:
```

### Sensor Names

The app senses and reports these values:

* `air_temp`: the air temperature in Celsius

* `air_humidity`: the relative humidity as a percent

* `air_pressure`: the air pressure in kPa

* `soil_moisture`: an indication of how much moisture is in the soil. A higher value means more moisture, a lower value means less moisture.

* `soil_temp`: the approximate soil temperature

* `light_level`: an indication of the ambient light level. A higher value means more ambient light, a lower value means less ambient light.

The sensor names are used both to report the values in monitoring and alert events, and as part of the environment variable names that you use to configure [alert thresholds](#alerts).

### Monitoring Events

The app sends a monitoring event to `greenhouse.qo` every `report_mins` minutes. The event contains
the readings of all sensors.

```json
{
      "air_humidity": 34.414062,
      "air_pressure": 1016.3697,
      "air_temp": 24.15,
      "app": "nf15",
      "light_level": 273,
      "soil_moisture": 344,
      "soil_temp": 26.489365
}
```

> When alert thresholds are configured, and one or more sensor readings is outside of the normal range, the event also includes `alert` and `alert_seq` properties to indicate the severity of the alert and the progress of the alert. See the section on [Alerts](#alerts) for details about these properties.

Reviewer 2: ```suggestion
When alert thresholds are configured, and one or more sensor readings are outside of the normal range, the event also includes `alert` and `alert_seq` properties to indicate the severity of the alert and the progress of the alert. See the section on [Alerts](#alerts) for details about these properties.
```

### Configuration Updates and Errors

When environment variables are changed, the app posts an event to `notify.qo`. The event contains details of which values changed and any configuration errors.

For example, setting the environment variable `environment_update_mins` to 1 produces this notification, indicating that the environment variable changed from the default of 5 minutes to 1 minute.

```json
{
    "updates": {
        "environment_update_mins": {
            "new_value": 1,
            "old_value": 5
        }
    }
}
```

Changing several variables, with some of them set incorrectly leads to both updates and errors being present:

```json
{
    "errors": {
        "environment_update_mins": {
            "error": "not a valid whole positive number.",
            "value": "1e"
        },
        "soil_temp_warning_high": {
            "error": "Value is not a number.",
            "value": "abc"
        }
    },
    "updates": {
        "light_level_normal_high": {
            "new_value": 2000
        },
        "light_level_normal_low": {
            "new_value": 20
        }
    }
}
```

## Alerts

> In addition to reporting the sensor values, the app can send an alert should a particular sensor value be too high or too low. The app uses environment variables to configure alert ranges for each sensor. The alert ranges are:

Reviewer 2: ```suggestion
In addition to reporting the sensor values, the app can send an alert when a particular sensor value is too high or too low. The app uses environment variables to configure alert ranges for each sensor. The alert ranges are:
```

Author: 'be' lol. Not sure I wrote that, I wasn't even wearing a pirate hat. ;-)

  * `first`: This is set when the first alert event is sent. The event is sent immediately.

  * `ongoing`: This is set when the alert condition is still present and the alert is ongoing. An event is sent every `report_mins` minutes for as long as the alert condition is still present.

  * `cleared`: This is set when the alert condition is no longer present and the alert is cleared. The event is sent immediately at the end of the alert.

 * `status`: Describes the status of a sensor value:

   * `ok`: The sensor value is in normal range, that is, above any configured low threshold and lower than any configured high threshold.

   * `low`: The sensor value is too low.

   * `high`: The sensor value is too high.

>**Note**: The `alert_seq` field makes it easy to detect when an alert starts and stops. You can use this to send external notifications immediately when an alert is detected and when it is cleared. Events with `alert_seq` set to `ongoing` provide periodic reminders that the alert is still ongoing.

### Soil Moisture and Light Level

> The `soil_moisture` and `light_level` are scalar readings that correlate to the amount of moisture in the soil and the amount of light, with more moisture and more light resulting is higher values, and less moisture and less light resulting in lower values. We suggest you use the app in the greenhouse for a few days, and use the monitoring data send to `greenhouse.qo` to determine the normal ranges for these values. You can then set the environment variables `soil_moisture_normal_low` etc.. to reflect the expected normal range of values.

Reviewer 2: ```suggestion
The `soil_moisture` and `light_level` are scalar readings that correlate to the amount of moisture in the soil and the amount of light, with more moisture and more light resulting in higher values, and less moisture and less light resulting in lower values. We suggest you use the app in the greenhouse for a few days, and use the monitoring data sent to `greenhouse.qo` to determine the normal ranges for these values. You can then set the environment variables `soil_moisture_normal_low` etc.. to reflect the expected normal range of values.
```
